profiling/src/exception.rs

```diff
@@ -0,0 +1,130 @@
+use crate::bindings as zend;
+use crate::PROFILER;
+use crate::REQUEST_LOCALS;
+use log::{error, trace};
+use std::cell::RefCell;
+use std::sync::atomic::AtomicI64;
+use std::sync::atomic::Ordering;
+
+use rand_distr::{Distribution, Poisson};
+
+/// The engine's previous throw exception hook
+static mut PREV_ZEND_THROW_EXCEPTION_HOOK: Option<zend::VmZendThrowExceptionHook> = None;
+
+/// take a sample every 100 exceptions
+pub static EXCEPTION_PROFILING_INTERVAL: AtomicI64 = AtomicI64::new(100);
+
+pub struct ExceptionProfilingStats {
+    /// number of exceptions until next sample collection
+    next_sample: i64,
+}
+
+impl ExceptionProfilingStats {
+    fn new() -> ExceptionProfilingStats {
+        ExceptionProfilingStats {
+            next_sample: ExceptionProfilingStats::next_sampling_interval(),
+        }
+    }
+
+    fn next_sampling_interval() -> i64 {
+        Poisson::new(EXCEPTION_PROFILING_INTERVAL.load(Ordering::Relaxed) as f64)
+            .unwrap()
+            .sample(&mut rand::thread_rng()) as i64
+    }
+
+    fn track_exception(&mut self, name: String) {
+        self.next_sample -= 1;
+
+        if self.next_sample > 0 {
+            return;
+        }
+
+        self.next_sample = ExceptionProfilingStats::next_sampling_interval();
+
+        REQUEST_LOCALS.with(|cell| {
+            // try to borrow and bail out if not successful
+            let locals = match cell.try_borrow() {
+                Ok(locals) => locals,
+                Err(_) => {
+                    return;
+                }
+            };
+
+            if let Some(profiler) = PROFILER.lock().unwrap().as_ref() {
+                // Safety: execute_data was provided by the engine, and the profiler doesn't mutate it.
+                unsafe {
+                    profiler.collect_exception(
+                        zend::ddog_php_prof_get_current_execute_data(),
+                        name,
+                        &locals,
+                    )
+                };
+            }
+        });
+    }
+}
+
+thread_local! {
+    static EXCEPTION_PROFILING_STATS: RefCell<ExceptionProfilingStats> = RefCell::new(ExceptionProfilingStats::new());
+}
+
+pub fn exception_profiling_minit() {
+    unsafe {
+        PREV_ZEND_THROW_EXCEPTION_HOOK = zend::zend_throw_exception_hook;
+        zend::zend_throw_exception_hook = Some(exception_profiling_throw_exception_hook);
```

Comment:
@realFlowControl I'm not really persuaded that the way this hook is used, is the right way.
After all, this hook is used for every single zend_throw_exception_internal.

What you are effectively sampling for is hence the number of frames plus the number of try+catch/finally constructs the exception falls through.

This will skew the exception profiling heavily towards exceptions which fall through more functions. I.e. an exception which is thrown and caught just in the function below is sampled much more seldom than an exception which comes 10 frames deep from within a library and then is caught by the caller.

To guard against that, I'd propose to:
a) take more often an exception sample
b) store the pointer to the last exception as well as the pointer to the last exception stacktrace [reduce pointer collision probability].

Reply:
> What you are effectively sampling for is hence the number of frames plus the number of try+catch/finally constructs the exception falls through.

As in:
```php
function first() {
    try {
        second();
    } catch (\TypeError $e) {
        // let the \RuntimeException bubble
    }
}

function second() {
    throw new \RuntimeException("foobar");
}

try {
    first();
} catch (\Exception $e) {
    // catched
}
```

would call the hook twice? As I do not observe this. This example PHP code will call my hook exactly once.

Reply:
It's 7.1 behaviour actually, I thought it's been that way for longer.

Reply:
I could reproduce this on PHP 7.1 and it is identifiable by doing a `NULL` check on the `exception` given to that hook. So I added this check and commit explaining why this exist.